core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -299,25 +296,8 @@ final class ZManaged[-R, +E, +A] private (reservation: ZIO[R, E, Reservation[R,
    * the passing of its value to the specified continuation function `k`,
    * followed by the effect that it returns.
    */
-  def flatMap[R1 <: R, E1 >: E, B](f0: A => ZManaged[R1, E1, B]): ZManaged[R1, E1, B] =
-    ZManaged[R1, E1, B] {
-      Ref.make[List[Exit[Any, Any] => ZIO[R1, Nothing, Any]]](Nil).map { finalizers =>
-        Reservation(
-          acquire = ZIO.uninterruptibleMask { restore =>
-            reserve
-              .flatMap(resR => finalizers.update(resR.release :: _) *> restore(resR.acquire))
-              .flatMap(r => f0(r).reserve)
-              .flatMap(resR1 => finalizers.update(resR1.release :: _) *> restore(resR1.acquire))
-          },
-          release = exitU =>
-            for {
-              fs    <- finalizers.get
-              exits <- ZIO.foreach(fs)(_(exitU).run)
-              _     <- ZIO.done(Exit.collectAll(exits).getOrElse(Exit.unit))
-            } yield ()
-        )
-      }
-    }
+  def flatMap[R1 <: R, E1 >: E, B](f: A => ZManaged[R1, E1, B]): ZManaged[R1, E1, B] =
+    ZManaged(self.zio.flatMap { case (_, a) => f(a).zio })
```

Comment:
Isn't the finaliser of the first effect important to keep here?

Reply:
That's a good question. The finalizers returned from the underlying `ZIO` are only used for early release.

I'm wondering if the correct thing to do is only allow the resulting ZIO to be released early, or should an early release of the resulting ZIO should also release the flatMapped ZIO.

Will think about this. Thank you for pointing this out!

Reply:
Ok, it's definitely right to compose the early release finalizers as well. Thanks @regiskuckaertz.

Reply:
Ha, I even did that without thinking for `ZManaged.foreach`.

Reply:
nice 👌